### PR TITLE
Add vmaas_sync

### PR DESCRIPTION
**Pull request by blayson**

### CODE REVIEW

File: base/database/setup.go [hunk 1]
```diff
@@ -1,5 +1,74 @@
...
+}
+
 func ConfigureDatabase() {
+	dbConfig := loadDbConfig()
+	sqlDb := openConn(dbConfig)
+	verifyConn(sqlDb)
+}
+
+func loadDbConfig() *PostgreSQLConfig {
+	config := PostgreSQLConfig{
```
Comment (jdobes) [inline base/database/setup.go:29]: Seems good for now, but we should also get values from clowder bit later.

Comment (blayson) [reply base/database/setup.go:29]: agreed, need to investigate how to do it in Go :D

File: Dockerfile [hunk 2]
```diff
@@ -20,6 +20,7 @@ ADD manager                     /vuln4shift/manager
 ADD pyxis                       /vuln4shift/pyxis
 ADD scripts                     /vuln4shift/scripts
 ADD test                        /vuln4shift/test
+ADD vmsync                        /vuln4shift/vmsync
```
Comment (epapbak) [inline Dockerfile:23]: please align =)

File: base/logging/logger.go [hunk 3]
```diff
@@ -27,7 +27,7 @@ func ParseLoglevel(level string) (logrus.Level, error) {
 	case "FATAL":
 		return logrus.FatalLevel, nil
 	default:
-		return logrus.InfoLevel, errors.New("Invalid loglevel given")
+		return logrus.InfoLevel, errors.New("invalid loglevel given")
```
Comment (epapbak) [inline base/logging/logger.go:30]: I'm curious, why this change?

File: base/utils/database.go [hunk 4]
```diff
@@ -4,27 +4,17 @@ import (
...
-	return value
-}
-
 func GetDbURL() string {
-	dbHost := getDbVar("POSTGRES_HOST")
-	dbPort := getDbVar("POSTGRES_PORT")
-	dbName := getDbVar("POSTGRES_DB")
-	dbUser := getDbVar("POSTGRES_USER")
-	dbPassword := getDbVar("POSTGRES_PASSWORD")
+	dbHost := GetEnv("POSTGRES_HOST", "localhost")
```
Comment (epapbak) [inline base/utils/database.go:13]: If we are introducing default values here, it should be `vuln4shift_database` and `5432` right?

File: go.mod [hunk 5]
```diff
@@ -1,12 +1,13 @@
 module app
 
-go 1.16
+go 1.18
```
Comment (epapbak) [inline go.mod:3]: since we are not using anything specific from this version, what do you think about leaving this change for another PR?

**Review by epapbak — COMMENTED** [review]
some minor stuff

File: base/logging/logger.go (hunk 3 above)
Comment (blayson) [reply base/logging/logger.go:30]: Error string should not be capitalized or end with punctuation mark 
https://github.com/golang/go/wiki/CodeReviewComments#error-strings

File: base/utils/database.go (hunk 4 above)
Comment (blayson) [reply base/utils/database.go:13]: oh, I have tested it locally and forgot to change, will fix

File: Dockerfile (hunk 2 above)
Comment (blayson) [reply Dockerfile:23]: Thats quite interesting, fixed :)

File: go.mod (hunk 5 above)
Comment (jdobes) [reply go.mod:3]: agree, also we might need to wait until we're able to compile with 1.18, https://catalog.redhat.com/software/containers/ubi8/go-toolset/5ce8713aac3db925c03774d1?container-tabs=overview is still on 1.16

Comment (jdobes) [reply go.mod:3]: - In RHEL/Fedora/UBI there are no 1.18 RPM packages yet, we'd need custom UBI-based build image and install go 1.18 binaries manually. Or wait on Fedora 36, which is going out later this month, should contain golang 1.18 packages and base the build image on Fedora
- Or try to replace UBI-golang build image with upstream https://hub.docker.com/_/golang

Comment (jdobes) [reply go.mod:3]: The second option could be quite good, unless there will be any problem/incompatibility

Comment (jdobes) [reply go.mod:3]: And if we switch to 1.18 let's do the change in separate PR, which can be merged before this one

File: go.mod [hunk 6]
```diff
@@ -15,3 +16,49 @@ require (
 	gorm.io/driver/postgres v1.3.1
 	gorm.io/gorm v1.23.2
 )
+
+require (
```
Comment (jdobes) [inline go.mod:20]: Where does these indirect deps come from? Are there all deps from go.sum? Is this generated? Do these indirect deps need to be updated manually?

Comment (blayson) [reply go.mod:20]: yeah this were generated by `go mod tidy` command

File: go.mod (hunk 5 above)
Comment (blayson) [reply go.mod:3]: seems, like simple replace of builder image works

Comment (blayson) [reply go.mod:3]: @jdobes #23

File: go.mod (hunk 6 above)
Comment (jdobes) [reply go.mod:19]: found the reason - https://stackoverflow.com/questions/69432877/why-there-are-two-require-blocks-in-go-mod-since-go-1-17?noredirect=1&lq=1

previously, the go.mod was generated on Fedora, where is still golang 1.16, 1.18 will be in Fedora 36 later this month

File: vmsync/api.go [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+			CveList:          []string{".*"},
+			PageSize:         PageSize,
+			RhOnly:           true,
+			ErrataAssociated: true,
+		}
+		vmaasResponse := APICveResponse{}
+
+		statusCode, err := client.RetryRequest(http.MethodPost, VmaasCvesURL, &vmaasRequest, &vmaasResponse)
+		if err != nil {
+			logger.Warningf("Request %s %s failed: statusCode=%d, err=%s", http.MethodPost, VmaasCvesURL, statusCode, err)
```
Comment (epapbak) [inline vmsync/api.go:71]: I would use logger.Errorf since when this happens, we stop the execution. WDYT?

File: vmsync/vmsync.go [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+	})
+}
+
+func syncCveMetadata() {
+	apiCveMap, err := getAPICves()
+
+	if err != nil {
+		logger.Fatalf("Unable to get CVEs from VMaaS: %s", err)
+	}
+	logger.Infof("CVEs in VMaaS: %d", len(apiCveMap))
```
Comment (epapbak) [inline vmsync/vmsync.go:41]: It basically says the same as the log[ just before](https://github.com/RedHatInsights/vuln4shift-backend/pull/5/files#diff-f03f754ac47dfa235847556581a3690e853ae0fc4e8659ffea4a01645cd45df7R87), in `getAPICves`

File: vmsync/vmsync.go [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+	toDeleteCves := make([]models.Cve, 0, len(dbCveMap))
+	for _, dbCve := range dbCveMap {
+		toDeleteCves = append(toDeleteCves, dbCve)
+	}
+
+	logger.Infof("CVEs to sync: %d", len(toSyncCves))
+	logger.Infof("CVEs to delete: %d", len(toDeleteCves))
+
+	if err = syncCves(toSyncCves, toDeleteCves); err != nil {
+		logger.Infof("Error during syncing CVEs into database: %s", err)
```
Comment (epapbak) [inline vmsync/vmsync.go:82]: `logger.Errorf`, maybe?

**Review by epapbak — COMMENTED** [review]
Small comments. Aside from that, LGTM

File: base/models/utils.go [hunk 10]
```diff
@@ -3,10 +3,13 @@ package models
...
 	"gorm.io/gorm"
+	"gorm.io/gorm/logger"
 )
 
-// GetGormConnection creates gorm database śtruct
+// GetGormConnection creates gorm database struct
 func GetGormConnection(dsn string) (*gorm.DB, error) {
-	db, err := gorm.Open(postgres.Open(dsn))
+	db, err := gorm.Open(postgres.Open(dsn), &gorm.Config{
+		Logger: logger.Default.LogMode(logger.Error),
```
Comment (jdobes) [inline base/models/utils.go:12]: What is this logging when added? Why logger.Error?

File: conf/vmsync.env [hunk 11]
```diff
@@ -0,0 +1,6 @@
+POSTGRES_USER=vmaas_gatherer
+POSTGRES_PASSWORD=vmaas_gatherer_pwd
+LOGGING_LEVEL=INFO
+API_RETRIES=3
+PAGE_SIZE=5000
+BASE_URL=http://localhost:8080/api/v3/cves
```
Comment (jdobes) [inline conf/vmsync.env:6]: I'd just set `https://console.redhat.com/api/vmaas/v3/cves` here, not relying on local vmaas

File: vmsync/api.go [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+			return cveMap, err
+		}
+
+		cveList, err := json.Marshal(&vmaasResponse.CveList)
```
Comment (jdobes) [inline vmsync/api.go:75]: Why marshall and then unmarshall? Why not just fill `cveMap` in `for`?

File: vmsync/api.go (hunk 7 above)
Comment (jdobes) [reply vmsync/api.go:71]: I'd keep Warning, pyxis gatherer api.go is using the same. If needed it can be then changed on both places at once.

File: vmsync/vmsync.go [hunk 13]
```diff
@@ -0,0 +1,163 @@
...
+
+	toSyncCves := make([]models.Cve, 0, len(apiCveMap))
+	for cveName, apiCve := range apiCveMap {
+		var severity models.Severity
+		err := severity.Scan(apiCve.Impact)
+		if err != nil {
+			severity = models.NotSet
+		}
+
+		cvss2Score, _ := strconv.ParseFloat(apiCve.Cvss2Score, 32)
```
Comment (jdobes) [inline vmsync/vmsync.go:51]: Better check the err

File: vmsync/vmsync.go (hunk 9 above)
Comment (jdobes) [reply vmsync/vmsync.go:82]: Perhaps Fatalf, because we don't want to continue from here and print the success msg below

File: vmsync/vmsync.go [hunk 14]
```diff
@@ -0,0 +1,163 @@
...
+	if len(toDeleteCves) > 0 {
+		if err := deleteCves(toDeleteCves, tx); err != nil {
+			return errors.Wrap(err, "Unable to delete cves in database")
+		}
+	}
+
+	return tx.Commit().Error
+}
+
+func deleteCves(toDeleteCves []models.Cve, tx *gorm.DB) error {
```
Comment (jdobes) [inline vmsync/vmsync.go:112]: Let's not delete CVEs yet.

- It shouldn't work this way - it shouldn't forcibly delete all deps from other tables (vuln-engine deletes only CVEs which are not on any systems)
- keep the function here, just don't call it for now
- there will be changes needed on both this side and on pyxis gatherer side to fix concurrency issues on the cve table (deadlocks during insert and update) and the CVE cache which is created in both components (and will contain invalid data when some CVE is deleted)

**Review by jdobes — CHANGES_REQUESTED** [review]

File: base/models/utils.go (hunk 10 above)
Comment (blayson) [reply base/models/utils.go:12]: That was done to debug the queries, but then I have leave that in Error mode, as if I got it correctly it is a default mode.

Comment (jdobes) [reply base/models/utils.go:12]: Ok, thanks

File: .gitignore [hunk 15]
```diff
@@ -15,3 +15,5 @@
 # vendor/
 
 .idea/
+
+vendor/
```
Comment (yungbender) [inline .gitignore:19]: newline pls

File: conf/vmsync.env (hunk 11 above)
Comment (yungbender) [reply conf/vmsync.env]: also newline pls

File: dbadmin/migrations/004_grant_insert_update_vmaas_gatherer_down.sql [hunk 16]
```diff
@@ -0,0 +1,4 @@
+REVOKE INSERT, UPDATE, DELETE ON cve IN SCHEMA public FROM vmaas_gatherer;
+REVOKE DELETE ON account_cve_cache IN SCHEMA public FROM vmaas_gatherer;
+REVOKE DELETE ON image_cve IN SCHEMA public FROMvmaas_gatherer;
+REVOKE DELETE ON cluster_cve_cache IN SCHEMA public FROM vmaas_gatherer;
```
Comment (yungbender) [inline dbadmin/migrations/004_grant_insert_update_vmaas_gatherer_down.sql:4]: newline pls

File: dbadmin/migrations/004_grant_insert_update_vmaas_gatherer_up.sql [hunk 17]
```diff
@@ -0,0 +1,4 @@
+GRANT SELECT, INSERT, UPDATE, DELETE ON cve TO vmaas_gatherer;
+GRANT SELECT, DELETE ON account_cve_cache TO vmaas_gatherer;
+GRANT SELECT, DELETE ON image_cve TO vmaas_gatherer;
+GRANT SELECT, DELETE ON cluster_cve_cache TO vmaas_gatherer;
```
Comment (yungbender) [inline dbadmin/migrations/004_grant_insert_update_vmaas_gatherer_up.sql:4]: newline pls

**Review by yungbender — CHANGES_REQUESTED** [review]

File: vmsync/vmsync.go (hunk 8 above)
Comment (blayson) [reply vmsync/vmsync.go]: deleted

File: vmsync/api.go (hunk 12 above)
Comment (blayson) [reply vmsync/api.go]: changed

File: vmsync/vmsync.go (hunk 13 above)
Comment (blayson) [reply vmsync/vmsync.go]: Yep, done.
If the error I'm setting `cvss2/3Score` to 0.0

File: vmsync/vmsync.go (hunk 14 above)
Comment (blayson) [reply vmsync/vmsync.go:117]: Got it.

**Comment by yungbender** [comment]
@blayson I have found a problem when running it, the sql migration scripts are named `*_down.sql` `*_up.sql` , this wont work, it needs to be named `*.down.sql` and `*.up.sql`

**Comment by blayson** [comment]
> @blayson I have found a problem when running it, the sql migration scripts are named `*_down.sql` `*_up.sql` , this wont work with the go-migrate, it needs to be named `*.down.sql` and `*.up.sql`. After I changed it locally it works and syncs correctly 👍

Yeah, thanks, renamed

**Comment by jdobes** [comment]
removing these checks from .golangci.yaml, is it intentional? If yes, why?

**Comment by blayson** [comment]
> removing these checks from .golangci.yaml, is it intentional? If yes, why?

There was a warning that this checks are disabled in go1.18. So, I have decided to remove it from golangci.yaml as it has no effect.

**Review by jdobes — APPROVED** [review]
